src/Symfony/Bundle/FrameworkBundle/Routing/Loader/YamlFileLoader.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Routing\Loader;
+
+use Symfony\Bundle\FrameworkBundle\Controller\RedirectController;
+use Symfony\Bundle\FrameworkBundle\Controller\TemplateController;
+use Symfony\Component\Routing\Loader\YamlFileLoader as BaseYamlFileLoader;
+use Symfony\Component\Routing\RouteCollection;
+
+/**
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+class YamlFileLoader extends BaseYamlFileLoader
+{
+    private static $availableKeys = [
+        'template' => ['template', 'max_age', 'shared_max_age', 'private'],
+        'redirect' => ['redirect_to_route', 'permanent', 'ignore_attributes', 'keep_request_method', 'keep_query_params'],
+        'url_redirect' => ['redirect_to_url', 'permanent', 'scheme', 'http_port', 'https_port', 'keep_request_method'],
+    ];
+
+    protected function validate($config, $name, $path)
+    {
+        // remove invalid keys for values normalized below
+        foreach (array_unique(array_merge(...array_values(self::$availableKeys))) as $key) {
```

Comment:
can we validate the attributes per type of route?
can we validate that the routes using one and only one of the 3 possibles types?

Reply:
When I look at how the parent is implemented and extended, the only valuable thing I can think of checking here is about defining both "controller" and one of the three types we introduce.

Reply:
We shouldn't need such check since the xsd is already doing the job.

Reply:
I see now that we could do both what you said and enforcing not using the controller key as I proposed. I'll add the proper tests to cover it in Yaml; the XSD is already doing the job, and the PHP fluent interface does not allow invalid calls by design.

Reply:
Done in #35653. Also added a new trait to prevent calling `controller` with configurators.